packages/front-end/components/Settings/EditDataSource/types.d.ts

```diff
@@ -0,0 +1,19 @@
+import { DataSourceInterfaceWithParams } from "back-end/types/datasource";
+
+export type DataSourceUIMode = "edit" | "view" | "add";
+
+export type DataSourceEditingResourceType =
+  // id: dataSource.settings.queries[x].userIdTypes[y].userIdType
+  | "identifier_types"
+  // id: dataSource.settings.queries[x].exposure[y].id
+  | "experiment_assignment"
+  // id: dataSource.settings.queries[x].identityJoins[y].ids
+  | "identifier_join"
+  // id: dataSource.settings.notebookRunQuery
+  | "jupyter_notebook";
+
+export type DataSourceQueryEditingModalBaseProps = {
+  dataSource: DataSourceInterfaceWithParams;
+  onSave: (dataSource: DataSourceInterfaceWithParams) => void;
+  onCancel: () => void;
+};
```

Comment:
The interface for every query editing section must return the new data source when calling `onSave` and provide the data source and the `onCancel` function to close the modal.